### PR TITLE
colormapping.c: optimize kmeans for acquiring clusters

### PR DESCRIPTION
**Pull request by ralfbrown**
TL;DR: Yields a 30-fold speedup with two threads and 200-fold speedup with 32 when clicking on one of the "acquire" buttons.  Results are now perceptually instantaneous.

Found this while looking to remove additional SSE code.  The SSE random-number generator in common/points.h is only used by the color  mapping and (deprecated) color transfer modules, so I took a closer look at color mapping and found performance issues.  Switching the sampling from selecting random coordinates for a pixel to doing a linear scan and randomly deciding whether or not to use the current pixel got a three-fold improvement from better cache locality.  Eliminating the cache contention and atomic writes by accumulating statistics per thread and then applying a reduction afterward got a way bigger boost.  Then I tried eliminating the random-number generation and just including all pixels instead of a sample and got an additional speedup.

```
Thr	Master	seq	perthr	nosample
1	~1300			230
2	~3300	~1100	199	119
32	~3800	~1150	21	18
```
Yes, the original code was much slower multi-threaded than single-threaded.  Such is the cost of writing to the same location for each pixel from multiple threads.

### CODE REVIEW

**Comment by ralfbrown** [comment]
Release notes:
- the "acquire clusters" operation in the color mapping module has been speed up by a factor of 30 to 200, making the results perceptually instantaneous on clicking the button

**Review by TurboGit — APPROVED** [review]
Thanks!
